File: R_scripts/SP-GIES.R
library(pcalg)
library(tictoc)
library(igraph)
source("../cupc/cuPC.R")

## This file contains functions for calling the SP-GIES algorithm which is a two-part algorithm that first calls
## cupc to estimate a skeleton using the PC algorithm. Then runs GIES from the pcalg algorithm to estimate the final
## graph given observational and interventional data

# Given file paths for the dataset, targets and target indices, run SP-GIES
# Optionally provide a path to a pre-generated skeleton
run_from_file_sp_gies <- function(dataset_path, target_path, target_index_path, save_path, threshold=0, skeleton_path=FALSE, save_pc=FALSE) {
    dataset <- read.table(dataset_path, sep=",", header=FALSE)
    targets <- read.table(target_path, sep=",", header=FALSE)
    targets.index <- read.table(target_index_path, sep=",", header=FALSE)
    print(dim(dataset))

    targets <- split(targets, 1:nrow(targets))
    targets <- lapply(targets, function(x) x[!is.na(x)])
    targets_init <- list(integer(0))
    targets <- append(targets_init, targets)
    targets.index <- unlist(targets.index)

    if (is.character(skeleton_path)) {
        skeleton <- read.table(skeleton_path, sep=",", header=FALSE)
        skeleton <- as(skeleton,"matrix")
        skeleton[abs(skeleton) < threshold] = 0
	skeleton[abs(skeleton) >= threshold] = 1
	skeleton <- as.data.frame(skeleton)
        skeleton <- skeleton == 0
        class(skeleton) <- "logical"
	sp_gies_from_skeleton(dataset, targets, targets.index, skeleton, save_path, save_pc)
    }
    else {
        sp_gies(dataset, targets, targets.index, save_path, save_pc)
    }
}

# Given dataset (matrix of size # samples x # nodes), targets (list),  targets.index (list of size #samples) run the SP-GIES algorithm
# and save the adjacency matrix in the save_path. Also saves the adjacency matrix of the skeleton (output of the cupc algorithm)
# if save_pc is set to TRUE
# Also prints the time to solution which includes calculating the sufficient statistics for the PC algorithm
sp_gies <- function(dataset, targets, targets.index, save_path, save_pc=FALSE, max_degree=integer(0)) {
    tic("cupc")
    tic("gies")
    corrolationMatrix <- cor(dataset)
    p <- ncol(dataset)
    suffStat <- list(C = corrolationMatrix, n = nrow(dataset))
    cuPC_fit <- cu_pc(suffStat, p=p, alpha=0.01)
    print("The total time consumed by cuPC is:")
    toc()

    # TODO we don't want this included in time to solution
    if (save_pc) {
        write.csv(as(cuPC_fit@graph, "matrix") ,row.names = FALSE, file = paste(save_path, 'cupc-adj_mat.csv',sep = ''))
    }


    fixedGaps <- as(cuPC_fit@graph,"matrix")
    fixedGaps <- as.data.frame(fixedGaps)
    fixedGaps <- fixedGaps == 0
    class(fixedGaps) <- "logical"

    score <- new("GaussL0penIntScore", data = dataset, targets=targets, target.index=targets.index)
    result <- pcalg::gies(score, fixedGaps=fixedGaps, targets=targets, maxDegree=max_degree)
    print("The total time consumed by SP-GIES is:")
    toc()
    print(max(degree(graph_from_adjacency_matrix(result$repr$weight.mat(), weighted=TRUE))))
    write.csv(result$repr$weight.mat() ,row.names = FALSE, file = paste(save_path, 'sp-gies-adj_mat.csv',sep = ''))
 }

# Same method as above, except skeleton comes from another algorithm. fixedGaps is a logical array that
# is  FALSE for all edges in the skeleton and TRUE otherwise
 sp_gies_from_skeleton <- function(dataset, targets, targets.index, fixedGaps, save_path, save_pc=FALSE) {
    tic()
    score <- new("GaussL0penIntScore", data = dataset, targets=targets, target.index=targets.index)
    result <- pcalg::gies(score, fixedGaps=fixedGaps, targets=targets)
    print("The total time consumed by SP-GIES is:")
    toc()
    write.csv(result$repr$weight.mat() ,row.names = FALSE, file = paste(save_path, 'sp-gies-adj_mat.csv',sep = ''))
 }

File: Python_scripts/sp_gies.py
import pandas as pd
import numpy as np
import rpy2.robjects as robjects
from rpy2.robjects.packages import importr
import rpy2.robjects as ro
import rpy2.robjects.numpy2ri
from utils import edge_to_dag, adj_to_edge, adj_to_dag, get_scores
from rpy2.robjects.packages import SignatureTranslatedAnonymousPackage
rpy2.robjects.numpy2ri.activate()
pcalg = importr('pcalg')
base = importr('base')

def cu_pc(data, outdir):
    with open("cupc/cuPC.R") as file:
        string = ''.join(file.readlines())
    cupc = SignatureTranslatedAnonymousPackage(string, "cupc")
    ro.r.assign("data", data)
    rcode = 'cor(data)'
    corMat = ro.r(rcode)
    ro.r.assign("corrolationMatrix", corMat)

    p = data.shape[1]
    ro.r.assign("p",p)

    rcode = 'list(C = corrolationMatrix, n = nrow(data))'
    suffStat = ro.r(rcode)
    ro.r.assign("suffStat", suffStat)

    cuPC_fit = cupc.cu_pc(ro.r['suffStat'],p=ro.r['p'],alpha=0.05)
    ro.r.assign("cuPC_fit", cuPC_fit)

    rcode = 'as(cuPC_fit@graph, "matrix")'
    skel = ro.r(rcode)
    ro.r.assign("skel", skel)

    rcode = "write.csv(skel ,row.names = FALSE, file = paste({}, 'cupc-adj_mat.csv',sep = ''))".format(outdir)
    return skel

def sp_gies(data, outdir, skel=None, target_map=None):
    if skel is None:
        obs_data = data.loc[data['target']==0]
        obs_data = obs_data.drop(columns=['target'])
        obs_data = obs_data.to_numpy()
        skel = cu_pc(obs_data, outdir)
    fixed_gaps = np.array((skel == 0), dtype=int)
    target_index = data.loc[:, 'target'].to_numpy()
    if target_map is not None:
        target_index = np.array([0 if i == 0 else target_map[i] + 1 for i in target_index])
    targets = np.unique(target_index)[1:]  # Remove 0 the observational target
    target_index += 1  # R indexes from 1
    data = data.drop(columns=['target']).to_numpy()

    nr, nc = data.shape
    D = ro.r.matrix(data, nrow=nr, ncol=nc)
    ro.r.assign("data", D)

    rcode = ','.join(str(int(i)) for i in targets)
    rcode = 'append(list(integer(0)), list({}))'.format(rcode)
    T = ro.r(rcode)
    ro.r.assign("targets", T)

    TI = ro.IntVector(target_index)
    ro.r.assign("target_index", TI)

    nr, nc = fixed_gaps.shape
    FG = ro.r.matrix(fixed_gaps, nrow=nr, ncol=nc)
    ro.r.assign("fixed_gaps", FG)

    score = ro.r.new('GaussL0penIntScore', ro.r['data'], ro.r['targets'], ro.r['target_index'])
    ro.r.assign("score", score)
    result = pcalg.gies(ro.r['score'], fixedGaps=ro.r['fixed_gaps'], targets=ro.r['targets'], maxDegree=10)
    ro.r.assign("result", result)

    rcode = 'result$repr$weight.mat()'
    adj_mat = ro.r(rcode)
    ro.r.assign("adj_mat", adj_mat)
    rcode = 'write.csv(adj_mat, row.names = FALSE,' \
            ' file = paste("{}", "sp-gies-adj_mat.csv", sep=""))'.format(outdir)
    ro.r(rcode)
    return adj_mat